docs/tutorials/go/build-an-ecommerce-app/index.md

```diff
@@ -13,14 +13,690 @@ last_update:
+Ensure a local Temporal Service is running (e.g. with [the Temporal CLI](https://github.com/temporalio/cli)) and that you can access the Temporal Web UI from port `8233`.
+These services are necessary for you to build and run this project.
+
+## Create the Shopping cart Workflow
+
+In a typical web app, a user's shopping cart would be stored as a row or document in a database.
+While you can store shopping carts in a separate database using Temporal, you have another option: you can represent a shopping cart as a [long-lived Workflow](https://temporal.io/blog/very-long-running-workflows). This way, you get the benefits of state tracking without needing to sync from external storage.
+
+In the Go SDK, a Temporal Workflow is a function that takes 2 parameters: a Temporal Workflow context `ctx` and an arbitrary `value`.
+Temporal can handle pausing and restarting the Workflow as needed -- a Workflow is a durable construct.
+Begin by mocking out a `workflow.go` with `CartItem` and `CartState` structs, and a main Workflow function called `CartWorkflow`.
+
+```go
+package app
+
+import (
+	"go.temporal.io/sdk/workflow"
+)
+
+type (
+	CartItem struct {
+		ProductId int
+		Quantity  int
+	}
+
+	CartState struct {
+		Items []CartItem
+		Email string
+	}
+)
+
+func CartWorkflow(ctx workflow.Context, state CartState) error {
+	logger := workflow.GetLogger(ctx)
+}
+```
+
+This will be your starting point. Next, you'll add some features to the `CartWorkflow`.
+
+### Add Signal Handling for your cart
+
+To support adding and removing elements from the cart, the Workflow needs to respond to different types of [Signals](https://docs.temporal.io/encyclopedia/workflow-message-passing).
+Signals are a way to notify Workflows of external events.
+In the Go SDK, you use Signal channels to listen for messages that either add or remove items from a shopping cart.
+Add these to within your `CartWorkflow` block:
+
+```go
+channel := workflow.GetSignalChannel(ctx, "cartMessages")
+selector := workflow.NewSelector(ctx)
+
+selector.AddReceive(channel, func(c workflow.ReceiveChannel, _ bool) {
+	var signal interface{}
+	c.Receive(ctx, &signal)
+
+	var routeSignal RouteSignal
+	err := mapstructure.Decode(signal, &routeSignal)
+	if err != nil {
+		logger.Error("Invalid signal type %v", err)
+		return
+	}
+
+	switch {
+	case routeSignal.Route == RouteTypes.ADD_TO_CART:
+		var message AddToCartSignal
+		err := mapstructure.Decode(signal, &message)
+		if err != nil {
+			logger.Error("Invalid signal type %v", err)
+			return
+		}
+		AddToCart(&state, message.Item)
+
+	case routeSignal.Route == RouteTypes.REMOVE_FROM_CART:
+		var message RemoveFromCartSignal
+		err := mapstructure.Decode(signal, &message)
+		if err != nil {
+			logger.Error("Invalid signal type %v", err)
+			return
+		}
+		RemoveFromCart(&state, message.Item)
+})
+
+for {
+	selector.Select(ctx)
+}
```

Comment:
I'd personally want some more explanation of the code - routeSignal, GetSignalChannel etc.